### PR TITLE
HIVE-27489: HPL/SQL does not support table aliases on column names in loops.

### PR DESCRIPTION
**Pull request by mdayakar**
### What changes were proposed in this pull request?
For supporting Table type feature for HPL/SQL as a part of [HIVE-24383](https://issues.apache.org/jira/browse/HIVE-24383)  to access table attributes like FIRST, LAST,.. **expr_dot** section is introduced in _Hplsql.g4_ file which got impacted _<table-name>.<column-name>_ expression. So currently as a part of _expr_dot_ section parsing, if table attribute is not used then considering it as column name and using the same instead of using _<table-name>.<column-name>_.

### Why are the changes needed?
To support _table-name.column-name_ or _table-alias-name.column-name_ in the cursors(loops).

### Does this PR introduce _any_ user-facing change?
No

### Is the change a dependency upgrade?
No

### How was this patch tested?
Test cases are added.
_mvn test -Dtest=TestHplSqlViaBeeLine -pl itests/hive-unit -Pitests_

### CODE REVIEW

File: hplsql/src/main/java/org/apache/hive/hplsql/objects/TableClass.java [hunk 1]
```diff
@@ -136,4 +155,8 @@ public MethodDictionary methodDictionary() {
   public boolean rowType() {
     return rowType;
   }
+
+  public static boolean isTableAttributeExists(String property) {
+    return TABLE_ATTRIBUTES_LIST.contains(property);
```
Comment (kasakrisz) [inline hplsql/src/main/java/org/apache/hive/hplsql/objects/TableClass.java:160]: TABLE_ATTRIBUTES_LIST is more like a Set<String>.

Can `methodDictionary.containsKey(property)` be used instead?

Comment (mdayakar) [reply hplsql/src/main/java/org/apache/hive/hplsql/objects/TableClass.java]: Changed the code accordingly.

File: itests/hive-unit/src/test/java/org/apache/hive/beeline/TestHplSqlViaBeeLine.java [hunk 2]
```diff
@@ -610,6 +610,25 @@ public void testNullCast() throws Throwable {
...
+  public void testTableAliasInColumnName() throws Throwable {
+    String SCRIPT_TEXT =
+        "DROP TABLE IF EXISTS input;\n" +
+        "DROP TABLE IF EXISTS result;\n" +
+        "CREATE TABLE input (col1 string, col2 int);\n" +
+        "CREATE TABLE result (res string);\n" +
+        "INSERT INTO input VALUES('Hive', 3);\n" +
+        "CREATE PROCEDURE p1() AS\n" +
+        "  BEGIN\n" +
+        "    FOR rec IN (select tab.col1, tab.col2 num from input tab) LOOP\n" +
```
Comment (amansinha100) [inline itests/hive-unit/src/test/java/org/apache/hive/beeline/TestHplSqlViaBeeLine.java:623]: Will this fail if 2 columns have the same alias name defined and within the LOOP we try to reference it ?  It would be good to have a meaningful error such as 'ambiguous column reference'.   
If this is not supported in the current code, it is ok to file a separate Jira and go ahead with this.

**Review by amansinha100 — COMMENTED** [review]
One comment below.  Overall LGTM.

**Review by kasakrisz — APPROVED** [review]
+1 pending tests

File: itests/hive-unit/src/test/java/org/apache/hive/beeline/TestHplSqlViaBeeLine.java (hunk 2 above)
Comment (mdayakar) [reply itests/hive-unit/src/test/java/org/apache/hive/beeline/TestHplSqlViaBeeLine.java:623]: It is just giving 'Unhandled exception in HPL/SQL' and in HS2 log file also no useful information. I will raise a separate issue to handle it.

File: hplsql/src/main/java/org/apache/hive/hplsql/objects/TableClass.java [hunk 3]
```diff
@@ -96,9 +96,9 @@ public class TableClass implements HplClass {
       } else { // single column
         Row row = new Row();
         row.addColumn(
-                self.hplClass().columns().get(0).columnName(),
-                self.hplClass().columns.get(0).columnTypeString(),
-                args.get(1));
```
Comment (kasakrisz) [inline hplsql/src/main/java/org/apache/hive/hplsql/objects/TableClass.java:101]: Seems to be an unrelated change.

File: itests/hive-unit/pom.xml [hunk 4]
```diff
@@ -494,6 +494,12 @@
       <artifactId>postgresql</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>net.sf.supercsv</groupId>
+      <artifactId>super-csv</artifactId>
+      <version>${super-csv.version}</version>
+      <scope>test</scope>
+    </dependency>
```
Comment (kasakrisz) [inline itests/hive-unit/pom.xml:502]: Is this a necessary change?

Comment (mdayakar) [reply itests/hive-unit/pom.xml:502]: I was getting below error without this dependency so added above dependency.
```
[INFO] Running org.apache.hive.beeline.TestHplSqlViaBeeLine
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 27.295 s <<< FAILURE! - in org.apache.hive.beeline.TestHplSqlViaBeeLine
[ERROR] org.apache.hive.beeline.TestHplSqlViaBeeLine.testTableAliasInColumnName  Time elapsed: 0.31 s  <<< ERROR!
java.lang.NoClassDefFoundError: org/supercsv/encoder/CsvEncoder
        at org.apache.hive.beeline.BeeLine.<init>(BeeLine.java:186)
        at org.apache.hive.beeline.BeeLine.<init>(BeeLine.java:566)
        at org.apache.hive.beeline.TestBeeLineWithArgs.testCommandLineScript(TestBeeLineWithArgs.java:164)
        at org.apache.hive.beeline.TestHplSqlViaBeeLine.testScriptFile(TestHplSqlViaBeeLine.java:653)
        at org.apache.hive.beeline.TestHplSqlViaBeeLine.testScriptFile(TestHplSqlViaBeeLine.java:640)
        at org.apache.hive.beeline.TestHplSqlViaBeeLine.testTableAliasInColumnName(TestHplSqlViaBeeLine.java:629)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
        at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
        at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
        at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
        at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
        at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
        at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:377)
        at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:138)
        at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:465)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:451)
Caused by: java.lang.ClassNotFoundException: org.supercsv.encoder.CsvEncoder
        at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
        ... 36 more

```

File: hplsql/src/main/java/org/apache/hive/hplsql/objects/TableClass.java (hunk 3 above)
Comment (mdayakar) [reply hplsql/src/main/java/org/apache/hive/hplsql/objects/TableClass.java:101]: Actually there is no change in this class. As a part of initial solution some changes were in this class later they are reverted. It looks this change some format change.
